File: Python/bubble sort.py
def bubble_sort(arr):
    """
    Bubble Sort Algorithm:
    - Repeatedly steps through the list, compares adjacent elements,
      and swaps them if they're in the wrong order.
    - Time Complexity: O(n^2)
    """
    n = len(arr)

    # Traverse the array n times
    for i in range(n):
        # Inner loop to compare adjacent elements
        for j in range(0, n - i - 1):
            # If current element is greater than next, swap
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]

# ✅ Example
nums = [64, 34, 25, 12, 22]
bubble_sort(nums)
print("Bubble Sorted:", nums)  # Output: [12, 22, 25, 34, 64]

File: C++/bubble sort.cpp
#include <iostream>
#include <vector>
using namespace std;

void bubbleSort(vector<int>& nums) {
    int n = nums.size();
    // Perform n-1 passes
    for (int i = 0; i < n - 1; i++) {
        // Push largest element to end in each pass
        for (int j = 0; j < n - i - 1; j++) {
            if (nums[j] > nums[j + 1])
                swap(nums[j], nums[j + 1]);  // Swap if in wrong order
        }
    }
}

// ✅ Use-case
int main() {
    vector<int> nums = {5, 1, 4, 2, 8};
    bubbleSort(nums);
    for (int num : nums)
        cout << num << " ";  // Output: 1 2 4 5 8
}
